### ISSUE TITLE
High CPU usage

### ISSUE THREAD

**Issue by JohnAnon9771** [Shopify/ruby-lsp#1062]
### Ruby version

3.0.6

### Description

High CPU usage.

I noticed this issue today when quickly correcting an error reported by YARP, typically using copilot's autocomplete. Ruby starts consuming a substantial amount of CPU for an extended period, ultimately causing the Ruby LSP to crash completely. Here's a video demonstrating this behavior:

https://github.com/Shopify/ruby-lsp/assets/42195321/aa3892fc-8fe2-4001-9b1c-500cc9b06eaa

### Expected output

Should have detected the correction change quickly.

**Comment by vinistock** [comment]
Thank you for the bug report! At what point does the LSP crash? Until the end of the video it looks like it's still reporting diagnostics - and the process is still there at the end.

I haven't seen this when using the completions, but it'd be interesting to understand if Copilot is pasting the suggestion as a single `TextEdit` or multiple (which could be the source of the spike in CPU usage).

**Comment by JohnAnon9771** [comment]
> Thank you for the bug report! At what point does the LSP crash? Until the end of the video it looks like it's still reporting diagnostics - and the process is still there at the end.
> 
> I haven't seen this when using the completions, but it'd be interesting to understand if Copilot is pasting the suggestion as a single `TextEdit` or multiple (which could be the source of the spike in CPU usage).

In the video, I demonstrated when Ruby LSP stops reporting new diagnoses. Why does it get stuck on the 'yarp' error? When the error reported by 'yarp' appears, I try to add some blank spaces at the bottom to see if it changes the diagnosis (in this case, it should show some RuboCop violations), but nothing happens. It remains stuck on the 'yarp' error. Only after some time, the error alert goes away, and I go back to adding blanks, and the RuboCop diagnostics reappear.

time: 0.26 seconds

**Comment by JohnAnon9771** [comment]
> I haven't seen this when using the completions, but it'd be interesting to understand if Copilot is pasting the suggestion as a single `TextEdit` or multiple (which could be the source of the spike in CPU usage).

Today I also experienced this error without autocomplete... but unfortunately I didn't record it.

**Comment by JohnAnon9771** [comment]
> Yeah, the LSP must have gotten stuck doing something and couldn't fulfill other requests. But there's no crash in the sense of the process dying, otherwise VS Code would've thrown a bunch of dialogues saying that the server was crashing.
> 
> Are you able to consistently reproduce this? If so, you might be able to discover exactly what's taking too long by using this setting
> 
> ```js
> {
>   // Put this temporarily in your user settings
>   "ruby lsp.trace.server": "messages"
> }
> ```
> 
> When the LSP goes into the high CPU state, take a look at the `Output` tab and the `Ruby LSP` channel. It should be printing all requests made and their response times. If you can share which requests came in immediately before going into high CPU usages, that would really help diagnose.

Yes, I can reproduce this easily, I will analyze it later and send you the results. Thanks!

**Comment by vinistock** [comment]
Thank you very much for digging deeper into this. It looks like the completion request was the one taking too long.

I believe #1073 will mitigate this issue.

**Comment by JohnAnon9771** [comment]
> Thank you very much for digging deeper into this. It looks like the completion request was the one taking too long.
> 
> I believe #1073 will mitigate this issue.

I performed new tests after version 0.11.2, and the high CPU usage behavior persists, even with the PR #1073 fix.

**Comment by JohnAnon9771** [comment]
I will try to answer your questions.
> * Where in the file do you insert the letter and trigger the slowness? The nesting here is particularly important

I always perform these tests in the same way as shown in the video, modifying the `proc_for_binds` method in the 
`arel/nodes/homogeneous_in` file (Rails repository). This can help you test it on your machine.

> * Which key did you press? We only support completion for classes, modules and constants at the moment, so this has to be an uppercase letter

I always start with an uppercase letter, in this case, 'A'.

Every time I tested it, it was at this location and with the letter A. So far, I haven't been able to reproduce this issue with lowercase letters.

> Are you able to reproduce the slowness with a single keypress? There are two relevant parts for profiling this right:

I can reproduce it by pressing a key, but there is still quite a bit of noise in the log... I tried to make it as clean as possible to help, but I couldn't.
